libraries/AP_Compass/AP_Compass_QMC5883P.cpp

```diff
@@ -0,0 +1,241 @@
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Driver by RadioLink LjWang, Jun 2017
+ * GPS compass module See<http://www.radiolink.com>
+ */
+#include "AP_Compass_QMC5883P.h"
+
+#include <stdio.h>
+#include <utility>
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/utility/sparse-endian.h>
+#include <AP_Math/AP_Math.h>
+#include <GCS_MAVLink/GCS.h>
+
+
+//Register Address
+#define QMC5883P_REG_ID 0x00
+#define QMC5883P_REG_DATA_OUTPUT_X 0x01
+#define QMC5883P_REG_STATUS 0x09
+#define QMC5883P_REG_CONF1 0x0A
+#define QMC5883P_REG_CONF2 0x0B
+
+#define QMC5883P_ID_VAL 0x80
+//
+//Register values
+// Sensor operation modes
+#define QMC5883P_MODE_SUSPEND    0x00
+#define QMC5883P_MODE_NORMAL     0x01
+#define QMC5883P_MODE_SINGLE     0x02
+#define QMC5883P_MODE_CONTINUOUS 0x03
+//
+// ODR data output rates for 5883L
+#define QMC5883P_ODR_10HZ  (0x00 << 2)
+#define QMC5883P_ODR_50HZ  (0x01 << 2)
+#define QMC5883P_ODR_100HZ (0x02 << 2)
+#define QMC5883P_ODR_200HZ (0x03 << 2)
+//
+// Over sampling Ratio OSR1
+#define QMC5883P_OSR1_8 (0x00 << 4)
+#define QMC5883P_OSR1_4 (0x01 << 4)
+#define QMC5883P_OSR1_2	(0x02 << 4)
+#define QMC5883P_OSR1_1	(0x03 << 4)
+//
+// Down sampling Rate OSR2
+#define QMC5883P_OSR2_1 (0x00 << 6)
+#define QMC5883P_OSR2_2 (0x01 << 6)
+#define QMC5883P_OSR2_4	(0x02 << 6)
+#define QMC5883P_OSR2_8	(0x03 << 6)
+
+//RNG
+#define QMC5883P_RNG_30G (0x00 << 2)
+#define QMC5883P_RNG_12G (0x01 << 2)
+#define QMC5883P_RNG_8G  (0x10 << 2)
+#define QMC5883P_RNG_2G  (0x11 << 2)
+
+//Reset
+#define QMC5883P_RST 0x80
+
+extern const AP_HAL::HAL &hal;
+
+AP_Compass_Backend *AP_Compass_QMC5883P::probe(AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev,
+                                               bool force_external,
+                                               enum Rotation rotation)
+{
+	GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "Before Probe");
+	if (!dev) {
+        return nullptr;
+    }
+	GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "dev found");
+    AP_Compass_QMC5883P *sensor = new AP_Compass_QMC5883P(std::move(dev),force_external,rotation);
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+	GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "sensor found");
+    return sensor;
+}
+
+AP_Compass_QMC5883P::AP_Compass_QMC5883P(AP_HAL::OwnPtr<AP_HAL::Device> dev,
+                                         bool force_external,
+                                         enum Rotation rotation)
+    : _dev(std::move(dev))
+    , _rotation(rotation)
+	, _force_external(force_external)
+{
+}
+
+bool AP_Compass_QMC5883P::init()
+{
+    _dev->get_semaphore()->take_blocking();
+
+    _dev->set_retries(10);
+#if 0
+    _dump_registers();
+#endif
+    GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "Before Whoami");
+    if(!_check_whoami()){
+         GCS_SEND_TEXT(MAV_SEVERITY_ALERT, "Before Whoami Fail");
+    	 goto fail;
+    }
+
+    if (!_dev->write_register(0x0B, 0x01)||
```

Comment:
try to get whitespace right
for new files, use Tools/CodeStyle/ardupilot-astyle.sh